haskell/test/Test/SetPosition.hs

```diff
@@ -0,0 +1,841 @@
+          , sppToY = liquidityProvider
+          , sppDeadline = deadline
+          , sppMaximumTokensContributed = PerToken 1000000 1000000
+          }
+          & expectFailedWith invalidWitnessErr
+
+test_fails_if_its_past_the_deadline :: TestTree
+test_fails_if_its_past_the_deadline =
+  nettestScenarioOnEmulatorCaps "fails if it's past the deadline" $ do
+    let liquidityDelta = 10000000
+    let lowerTickIndex = -10
+    let upperTickIndex = 15
+
+    liquidityProvider <- newAddress auto
+    let xTokenId = FA2.TokenId 0
+    let yTokenId = FA2.TokenId 1
+    let xFa2storage = FA2.Storage
+          { sLedger = mkBigMap [ ((liquidityProvider, xTokenId), 100000) ]
+          , sOperators = mempty
+          , sTokenMetadata = mempty
+          }
+    let yFa2storage = FA2.Storage
+          { sLedger = mkBigMap [ ((liquidityProvider, yTokenId), 100000) ]
+          , sOperators = mempty
+          , sTokenMetadata = mempty
+          }
+    xToken <- originateSimple "fa2" xFa2storage (FA2.fa2Contract def { FA2.cAllowedTokenIds = [xTokenId] })
+    yToken <- originateSimple "fa2" yFa2storage (FA2.fa2Contract def { FA2.cAllowedTokenIds = [yTokenId] })
+
+    let initialSt = mkStorage xToken xTokenId yToken yTokenId
+    cfmm <- originateSegCFMM FA2 FA2 initialSt
+
+    withSender liquidityProvider do
+      call xToken (Call @"Update_operators") [FA2.AddOperator $ FA2.OperatorParam liquidityProvider (toAddress cfmm) xTokenId]
+      call yToken (Call @"Update_operators") [FA2.AddOperator $ FA2.OperatorParam liquidityProvider (toAddress cfmm) yTokenId]
+
+    withSender liquidityProvider do
+      now <- getNow
+      let expiredDeadline = now `timestampPlusSeconds` -1
+      call cfmm (Call @"Set_position")
+        SetPositionParam
+          { sppLowerTickIndex = lowerTickIndex
+          , sppUpperTickIndex = upperTickIndex
+          , sppLowerTickWitness = minTickIndex
+          , sppUpperTickWitness = minTickIndex
+          , sppLiquidityDelta = liquidityDelta
+          , sppToX = liquidityProvider
+          , sppToY = liquidityProvider
+          , sppDeadline = expiredDeadline
+          , sppMaximumTokensContributed = PerToken 1000000 1000000
+          }
+          & expectFailedWith pastDeadlineErr
+
+      let validDeadline = now
+      call cfmm (Call @"Set_position")
+        SetPositionParam
+          { sppLowerTickIndex = lowerTickIndex
+          , sppUpperTickIndex = upperTickIndex
+          , sppLowerTickWitness = minTickIndex
+          , sppUpperTickWitness = minTickIndex
+          , sppLiquidityDelta = liquidityDelta
+          , sppToX = liquidityProvider
+          , sppToY = liquidityProvider
+          , sppDeadline = validDeadline
+          , sppMaximumTokensContributed = PerToken 1000000 1000000
+          }
+
+test_cannot_set_position_over_max_tick :: TestTree
+test_cannot_set_position_over_max_tick =
+  nettestScenarioOnEmulatorCaps "cannot set a position with upper_tick > max_tick" $ do
```

Comment:
> Testing the lower_tick < min_tick scenario is left for future work

I didn't test that property explicitly because it's implied by these two other observations:
* You need a witness lower than or equal to your `lower_tick_index`
* The lowest witness available is `min_tick`

Therefore, it's impossible for `lower_tick` to be lower than `min_tick`.

The same observations do not, however, lead to the same guarantee about `upper_tick`, which is why I wrote an explicit test for that.

Reply:
Ah I see, right. Good then :+1: